File: source/cpp/953. Verifying an Alien Dictionary.cpp
#include <array>
#include <string>
#include <string_view>
#include <vector>


class Solution {
private:
    bool isSorted(std::string_view word1, std::string_view word2, const std::array<int, 26>& orderMap) const {
        for (auto i = 0; i < std::min(word1.size(), word2.size()); ++i) {
            if (word1[i] == word2[i]) continue;

            return orderMap[word1[i] - 'a'] < orderMap[word2[i] - 'a'];
        }

        return word1.size() <= word2.size();
    }
public:
    bool isAlienSorted(std::vector<std::string>& words, std::string_view order) const {
        std::array<int, 26> orderMap;
        for (auto i = 0; i < order.size(); ++i)
            orderMap[order[i] - 'a'] = i;

        for (int i = 0; i < words.size() - 1; ++i) {
            if (isSorted(words[i], words[i + 1], orderMap)) continue;
            
            return false;
        }

        return true;
    }
};

File: source/python/953. Verifying an Alien Dictionary.py
class Solution:
    def compare(self, word1: str, word2: str, order_dict: dict[str, int]) -> bool:
        for i in range(min(len(word1), len(word2))):
            if word1[i] == word2[i]:
                continue

            return order_dict[word1[i]] < order_dict[word2[i]]

        return len(word1) <= len(word2)

    def isAlienSorted(self, words: list[str], order: str) -> bool:
        order_dict = {c: i for i, c in enumerate(order)}

        for i in range(len(words) - 1):
            if not self.compare(words[i], words[i + 1], order_dict):
                return False

        return True


def test():
    s = Solution()

    assert s.isAlienSorted(
        ["hello", "leetcode"], "hlabcdefgijkmnopqrstuvwxyz") == True
    assert s.isAlienSorted(
        ["word", "world", "row"], "worldabcefghijkmnpqstuvxyz") == False
    assert s.isAlienSorted(
        ["apple", "app"], "abcdefghijklmnopqrstuvwxyz") == False

    print("All tests passed!")


if __name__ == "__main__":
    test()
